ycmd/completers/go/gocode_completer.py

```diff
@@ -120,6 +120,56 @@ def FindGoCodeBinary( self, user_options ):
     return utils.PathToFirstExistingExecutable( [ 'gocode' ] )
 
 
+  def DefinedSubcommands( self ):
+    return GoCodeCompleter.subcommands.keys()
+
+
+  def OnFileReadyToParse( self, request_data ):
+    self._StartServer()
+
+
+  def OnUserCommand( self, arguments, request_data ):
+    if not arguments:
+      raise ValueError( self.UserCommandsHelpMessage() )
+
+    command = arguments[ 0 ]
+    if command in GoCodeCompleter.subcommands:
+      command_lamba = GoCodeCompleter.subcommands[ command ]
+      return command_lamba( self, request_data )
+    else:
+      raise ValueError( self.UserCommandsHelpMessage() )
+
+
+  def Shutdown( self ):
+    self._StopServer()
+
+
+  def _StartServer( self ):
+    """ Start the GoCode server """
+    self._GoCode()
+
+
+  def _StopServer( self ):
+    """ Stop the GoCode server """
+    _logger.info( 'Stopping GoCode server' )
+    self._GoCode( 'close' )
```

Comment:
I'd wrap this call with a `_CloseGoCodeServer` function that in turn passes `close`.